idr/node.go

```diff
@@ -0,0 +1,124 @@
+package idr
+
+import (
+	"fmt"
+	"strings"
+)
+
+// NodeType is the type of a Node in an IDR.
+type NodeType uint
+
+const (
+	// DocumentNode is the type of the root Node in an IDR tree.
+	DocumentNode NodeType = iota
+	// ElementNode is the type of an element Node in an IDR tree.
+	ElementNode
+	// TextNode is the type of an text/data Node in an IDR tree.
+	TextNode
+	// AttributeNode is the type of an attribute Node in an IDR tree.
+	AttributeNode
+)
+
+// String converts NodeType to a string.
+func (nt NodeType) String() string {
+	switch nt {
+	case DocumentNode:
+		return "DocumentNode"
+	case ElementNode:
+		return "ElementNode"
+	case TextNode:
+		return "TextNode"
+	case AttributeNode:
+		return "AttributeNode"
+	default:
+		return fmt.Sprintf("(unknown NodeType: %d)", nt)
+	}
+}
+
+// Node represents a node of element/data in an IDR (intermediate data representation) ingested and created
+// by the omniparser.
+// Credit: this is by and large a copy and some adaptation from
+// https://github.com/antchfx/xmlquery/blob/master/node.go. The reasons we want to have our own struct:
+// - more stability
+// - one struct to represent XML/JSON/EDI/CSV/txt/etc. Vs antchfx's work have one struct (in each repo)
+//   for each format.
+// - Node allocation recycling.
+type Node struct {
+	Parent, FirstChild, LastChild, PrevSibling, NextSibling *Node
+
+	Type NodeType
+	Data string
+
+	FormatSpecific interface{}
+}
+
+// CreateNode creates a generic *Node.
+func CreateNode(ntype NodeType, data string) *Node {
+	return &Node{
+		Type: ntype,
+		Data: data,
+	}
+}
+
+// InnerText returns a Node's children's texts concatenated.
+// Note (in an XML IDR tree) none of the AttributeNode's text will be included.
```

Comment:
What's the reason that we can't include AttributeNode's text? I mean I can see `antchfx`'s one include it.

Reply:
antchfx and this IDR design is slightly different and I made the change purposefully. In antchfx design, attr is carried on Node directly
```
type Node struct {
    Attr []xml.Attr
}
```
There is no AttributeNode in the node tree constructed from input. Attributes are all directly carried in Attr field on ElementNode. That's antchfx design. Which is fine for xml node, but if we use the node type for other formats, it's really weird. And I don't like to treat Attr differently. So my design is to have AttributeNode as regular node - as explained in IDR's readme.md. The only catch is InnerText needs to exclude them - since antchfx's InnerText never include them neither - because AttributeNode ever exists from an XML parsed Node tree.